tests/test_smartcache_patch_wsi_dataset.py

```diff
@@ -0,0 +1,168 @@
+        {"image": np.array([[[246]], [[246]], [[246]]], dtype=np.uint8), "label": np.array([[[1]]])},
+        {"image": np.array([[[246]], [[246]], [[246]]], dtype=np.uint8), "label": np.array([[[1]]])},
+        {"image": np.array([[[247]], [[245]], [[248]]], dtype=np.uint8), "label": np.array([[[1]]])},
+        {"image": np.array([[[245]], [[247]], [[244]]], dtype=np.uint8), "label": np.array([[[1]]])},
+        {"image": np.array([[[246]], [[246]], [[246]]], dtype=np.uint8), "label": np.array([[[1]]])},
+        {"image": np.array([[[246]], [[246]], [[246]]], dtype=np.uint8), "label": np.array([[[1]]])},
+        {"image": np.array([[[247]], [[245]], [[248]]], dtype=np.uint8), "label": np.array([[[2]]])},
+        {"image": np.array([[[245]], [[247]], [[244]]], dtype=np.uint8), "label": np.array([[[2]]])},
+        {"image": np.array([[[246]], [[246]], [[246]]], dtype=np.uint8), "label": np.array([[[2]]])},
+        {"image": np.array([[[246]], [[246]], [[246]]], dtype=np.uint8), "label": np.array([[[2]]])},
+        {"image": np.array([[[247]], [[245]], [[248]]], dtype=np.uint8), "label": np.array([[[2]]])},
+        {"image": np.array([[[245]], [[247]], [[244]]], dtype=np.uint8), "label": np.array([[[2]]])},
+        {"image": np.array([[[246]], [[246]], [[246]]], dtype=np.uint8), "label": np.array([[[2]]])},
+        {"image": np.array([[[246]], [[246]], [[246]]], dtype=np.uint8), "label": np.array([[[2]]])},
+        {"image": np.array([[[247]], [[245]], [[248]]], dtype=np.uint8), "label": np.array([[[0]]])},
+        {"image": np.array([[[245]], [[247]], [[244]]], dtype=np.uint8), "label": np.array([[[0]]])},
+        {"image": np.array([[[246]], [[246]], [[246]]], dtype=np.uint8), "label": np.array([[[0]]])},
+        {"image": np.array([[[246]], [[246]], [[246]]], dtype=np.uint8), "label": np.array([[[0]]])},
+    ],
+]
+
+
+class TestSmartCachePatchWSIDataset(unittest.TestCase):
+    @parameterized.expand(
+        [
+            TEST_CASE_0,
+            TEST_CASE_1,
+            TEST_CASE_2,
+        ]
+    )
+    @skipUnless(has_cim, "Requires CuCIM")
+    def test_read_patches(self, file_url, input_parameters, expected):
+        self.camelyon_data_download(file_url)
+        dataset = SmartCachePatchWSIDataset(**input_parameters)
+        self.assertEqual(len(dataset), input_parameters["cache_num"])
+        total_num_samples = len(input_parameters["data"])
+        num_epochs = int(
+            np.ceil(total_num_samples / (input_parameters["cache_num"] * input_parameters["replace_rate"]))
+        )
+
+        dataset.start()
+        i = 0
+        for _ in range(num_epochs):
+            for j in range(len(dataset)):
+                samples = dataset[j]
+                n_patches = len(samples)
+                self.assert_samples_expected(samples, expected[i : i + n_patches])
+                i += n_patches
+            dataset.update_cache()
+        dataset.shutdown()
+
+    def camelyon_data_download(self, file_url):
```

Comment:
the method is implemented multiple times in the codebase, shall we move it into `tests/utils.py`? also, would be great to download into `tests/testing_data` folder

Reply:
Yeah that makes sense. I can do that.

Reply:
how is this resolved?

Reply:
@wyli, I'll do it later in another PR since it affects almost all the pathology related unittests.